File: client/src/test/scala/stasis/test/specs/unit/client/ops/integration/mocks/MockBackupLibraryKind.scala
package stasis.test.specs.unit.client.ops.integration.mocks

import java.time.Instant

import scala.concurrent.Future

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import stasis.client.collection.BackupCollector
import stasis.client.collection.rules.Rule
import stasis.client.collection.rules.SourceUri
import stasis.client.model.DatasetMetadata
import stasis.client.model.EntityMetadata
import stasis.client.model.EntityRef
import stasis.client.model.SourceEntity
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.backup.BackupEntityKind
import stasis.client.ops.backup.stages.EntityDiscovery
import stasis.client.ops.backup.{Providers => BackupProviders}
import stasis.shared.ops.Operation

class MockBackupLibraryKind(library: MockLibrary) extends BackupEntityKind.Library {
  override def scheme: String = library.scheme

  override def collector(
    collector: EntityDiscovery.Collector,
    latestMetadata: Option[DatasetMetadata],
    providers: BackupProviders,
    parallelism: ParallelismConfig
  )(implicit operation: Operation.Id, mat: Materializer): Future[BackupCollector] = {
    val rules = collector match {
      case EntityDiscovery.Collector.WithRules(rules) => rules
      case _                                          => Seq.empty
    }

    val libraryRules = rules.filter(rule => SourceUri.scheme(rule.source).contains(scheme))
    val includes = libraryRules.collect { case rule if rule.operation == Rule.Operation.Include => rule.source }
    val excludes = libraryRules.collect { case rule if rule.operation == Rule.Operation.Exclude => rule.source }

    val sourceEntities = library.entries.toList.collect {
      case (key, content) if includes.exists(key.startsWith) && !excludes.exists(key.startsWith) =>
        SourceEntity(
          ref = EntityRef.default(key),
          existingMetadata = None,
          currentMetadata = libraryMetadata(key, content)
        )
    }

    Future.successful(() => Source(sourceEntities))
  }

  override def read(entity: SourceEntity, ref: EntityRef.Library, chunkSize: Int): Source[ByteString, NotUsed] =
    Source.single(library.entries(ref.key))

  private def libraryMetadata(key: String, content: ByteString): EntityMetadata.Library =
    EntityMetadata.Library(
      path = key,
      created = Instant.now(),
      updated = Instant.now(),
      size = content.length.toLong,
      checksum = BigInt(content.length),
      crates = Map.empty,
      compression = "deflate",
      attributes = ByteString(s"attributes:$key")
    )
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/ops/integration/mocks/MockBackupLibraryKind.kt
package stasis.test.client_android.lib.ops.integration.mocks

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import okio.Buffer
import okio.ByteString
import okio.ByteString.Companion.encodeUtf8
import okio.Source
import stasis.client_android.lib.collection.BackupCollector
import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.collection.rules.SourceUri
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.EntityRef
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.backup.BackupEntityKind
import stasis.client_android.lib.ops.backup.Providers
import stasis.client_android.lib.ops.backup.stages.EntityDiscovery
import java.math.BigInteger
import java.time.Instant

class MockBackupLibraryKind(private val library: MockLibrary) : BackupEntityKind.Library {
    override val scheme: String = library.scheme

    override suspend fun collector(
        operation: OperationId,
        collector: EntityDiscovery.Collector,
        latestMetadata: DatasetMetadata?,
        providers: Providers
    ): BackupCollector {
        val rules = when (collector) {
            is EntityDiscovery.Collector.WithRules -> collector.rules
            else -> emptyList()
        }

        val libraryRules = rules.filter { SourceUri.scheme(it.source) == scheme }
        val includes = libraryRules.filter { it.operation == Rule.Operation.Include }.map { it.source }
        val excludes = libraryRules.filter { it.operation == Rule.Operation.Exclude }.map { it.source }

        val sourceEntities = library.entries
            .filterKeys { key -> includes.any { key.startsWith(it) } && excludes.none { key.startsWith(it) } }
            .map { (key, content) ->
                SourceEntity(
                    ref = EntityRef.default(key),
                    existingMetadata = null,
                    currentMetadata = libraryMetadata(key, content)
                )
            }

        return object : BackupCollector {
            override fun collect(): Flow<SourceEntity> = sourceEntities.asFlow()
        }
    }

    override fun read(entity: SourceEntity, ref: EntityRef.Library): Source =
        Buffer().write(library.entries.getValue(ref.key))

    private fun libraryMetadata(key: String, content: ByteString): EntityMetadata.Library =
        EntityMetadata.Library(
            path = key,
            created = Instant.now(),
            updated = Instant.now(),
            size = content.size.toLong(),
            checksum = BigInteger.valueOf(content.size.toLong()),
            crates = emptyMap(),
            compression = "deflate",
            attributes = "attributes:$key".encodeUtf8()
        )
}
